README.md
# AnimDessin2
This panel is designed for Photoshop CC 2014 to CC 2020+. It allows you to draw frame-by-image animations (cel-animation like), simplifying the process. it also lets you test the movie and set the duration of a keyframe…

### Short info
This [Photoshop Extension](https://helpx.adobe.com/photoshop/kb/plug-ins-extensions-photoshop-cc.html) was created by [Stéphane Baril](https://github.com/sbaril/Photoshop-Animation) aka sbaril with help of others. This repo is my add to this wonderful extension and i hope it helps others. I wanted to expand its functionality while learning the process of creating an Extension. 

I've added many updates and features which for the time being you can [read more about in this issue](https://github.com/sbaril/Photoshop-Animation/issues/8) on his repo


### Installing (manual method)
Since this extension is still a work in progress currently the only method to try it, i using the manual install method. Below are some steps the users needs to follow, its shows both Mac and Windows.

> Windows: Open regedit > HKEY_CURRENT_USER/Software/Adobe/CSXS.5, then add a new entry `PlayerDebugMode` of type “`string`” with the value of “`1`”.

> Mac: In the Terminal, type: `defaults write com.adobe.CSXS.5 PlayerDebugMode 1`

[Source > Adobe](https://github.com/Adobe-CEP/Getting-Started-guides/tree/d85c8e861ef90bb5c3443516d45c631e9fdebd7f/Client-side%20Debugging)

This is the correct info about the install location


**Debugging Unsigned Extensions**
If you are in the midst of development and are not using HTML Extension Builder for debug workflows and want to bypass the need to sign your extensions, you can bypass the check for extension signatures by editing the CSXS preference properties file, located at:

- Win: regedit > `HKEY_CURRENT_USER/Software/Adobe/CSXS.5`, then add a new entry` PlayerDebugMode` of type "string" with the value of "1".
- Mac: In the terminal, type: `defaults write com.adobe.CSXS.5 PlayerDebugMode `1 (The plist is also located at /Users/<username>/Library/Preferences/com.adobe.CSXS.8.plist)

These entries will enable debug extensions to be displayed in the host applications. Please note that, CSXS.8 is given with the assumption that, you are developing the extension for CEP 8. If you are developing extension for previous version of CEP, replace 8 with corresponding version number.

> Debugging is need to make the extension working without using the ZXP approach. These are signed and therefor run without blockage.

**Extension Folders**
CEP supports 3 types of extension folders.

Product extension folder. Here is a suggestion, but each point product can decide where this folder should be. Third party extension can't be installed at this location.

**System extension folder**
```Win(x86): C:\Program Files\Common Files\Adobe\CEP\extensions```
```Win(x64): C:\Program Files (x86)\Common Files\Adobe\CEP\extensions, and C:\Program Files\Common Files\Adobe\CEP\extensions (since CEP 6.1)```
```Mac: /Library/Application Support/Adobe/CEP/extensions```


**Per-user extension folder**

```Win: C:\Users\<USERNAME>\AppData\Roaming\Adobe\CEP/extensions```
```Mac: ~/Library/Application Support/Adobe/CEP/extensions```

[source > Adobe CEP](https://github.com/Adobe-CEP/CEP-Resources/blob/master/CEP_8.x/Documentation/CEP%208.0%20HTML%20Extension%20Cookbook.md#extension-folders
)



com.sbaril.animdessin2/CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.


## [0.8.1] - 2022-11-10
### Fixed
- Issue with InBetween After keeping 2 frames selected after new frame is added after > due to applyToAll, it reselect afterwards
- Panel height issue when toggling either Tooltips or GLobal Timeline OFF
- Tooltip for Show All Layers favorite button, corrected key-combo it showed

### Added
- New Doc Preset has updated localized labels 

## [0.0.1] - 2021-10-13
### Added
- Initial release

## Notes

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
<!--### Official Rigify Info-->


[0.8.1]:https://github.com/schroef/AnimDessin2/releases/tag/v0.8.1
[0.0.1]:https://github.com/schroef/AnimDessin2/releases/tag/v0.0.1

com.sbaril.animdessin2/js/toolguide.js
/*jslint vars: true, plusplus: true, devel: true, nomen: true, regexp: true, indent: 4, maxerr: 50 */
/*global $, window, location, CSInterface, SystemPath, themeManager*/

(function () {
    'use strict';

    var csInterface = new CSInterface();
    var gExtensionId = "com.sbaril.animdessin2.Guides";

    function Persistent(inOn) { // keep session stored

        if (inOn) {
            var event = new CSEvent("com.adobe.PhotoshopPersistent", "APPLICATION");
        } else {
            var event = new CSEvent("com.adobe.PhotoshopUnPersistent", "APPLICATION");
        }

        event.extensionId = gExtensionId;
        csInterface.dispatchEvent(event);
    }
    // Localize Languages
    function getLocalize() {
        var csInterface = new CSInterface();
        var renderBundle = csInterface.initResourceBundle();
        // console.log(renderBundle)
        return renderBundle
    }

    // Opens the chrome developer tools in host app
    // Depricated https://community.adobe.com/t5/labs/showdevtools-method-added-by-default/td-p/6905305?page=1
    function showDevTools() {
        window.__adobe_cep__.showDevTools();
    }

    // Reloads extension panel
    function reloadPanel() {
        location.reload();
    }

    // Loads / executes a jsx file
    function loadJSXFile(pPath) {
        var scriptPath = csInterface.getSystemPath(SystemPath.EXTENSION) + pPath;
        csInterface.evalScript('evalFile("' + scriptPath + '")');
    }

    function checkSettings() {
        toggleGuide();
    }
    function callToolMock() {
        window.location.href = "./toolsguide.html";
    }

    function callVideoMock() {
        window.location.href = "./videoguide.html";
    }
    // Show helpguide
    function toggleGuide(item) {
        if (item == "guide") {
            // else {
            localStorage.setItem('toggleGuide', "guideInActive");
            localStorage.setItem('toggleVideo', "videoInActive");
            $("#guide").addClass("InActive").removeClass("Active");
            $("#video").addClass("InActive").removeClass("Active");
            // document.getElementById("toggleVideo").Checked;
            // }
            if (localStorage.getItem('toggleGuide') == "guideInActive") {
                localStorage.setItem('toggleGuide', "guideActive");
                localStorage.setItem('toggleVideo', "videoInActive");
                $("#guide").addClass("Active").removeClass("InActive");
                $("#video").addClass("InActive").removeClass("Active");
                csInterface.resizeContent(320, 500);
                csInterface.updateContextMenuItem("toggleGuide", true, true);
                csInterface.updateContextMenuItem("toggleVideo", true, false);

            }
        }
        if (item == "close") {
            csInterface.closeExtension();
        }
        if (item == "videoMock") {
            csInterface.resizeContent(320, 500);
            callVideoMock();
        }
        if (item == "toolMock") {
            csInterface.resizeContent(320, 500);
            callToolMock();
        }
    }

    function sendWarning() {
        loadJSXFile("/jsx/AnimD2_warning.jsx");
    }

    function init() {
        themeManager.init();

        //Persistent(true); //persistent to prevent extension from unloading


        ////////////////////////////////////////////////////////////////////////////////
        // ANIMDESSIN2 TOOLGUIDE BUTTONS FUNCTIONS
        ////////////////////////////////////////////////////////////////////////////////   

        $("#btn_debug").click(showDevTools);
        $("#btn_reload").click(reloadPanel);

        $("#btn_closeGuide").click(toggleGuide);
        $("#btn_closeVideo").click(toggleGuide);

        if (navigator.platform === "MacIntel") $("#newDocSet").addClass("kcmb-sh-al-osx-lm");
        if (navigator.platform === "Win32") $("#newDocSet").addClass("kcmb-sh-al-win-lm");

        ////////////////////////////////////////////////////////////////////////////////
        // ANIMDESSIN TOOLGUIDE FLYOUT MENU
        ////////////////////////////////////////////////////////////////////////////////

        // Ugly workaround to keep track of "checked" and "enabled" statuses
        var checkableMenuItem_isChecked = true;
        var targetMenuItem_isEnabled = true;


        function flyoutXML() {
            return (
            '<Menu> \n<MenuItem Id="AboutItemAnimD2" Label="' +
                getLocalize().flyout_about +
                '"/> \n<MenuItem Id="OpenWebsiteAnimD2" Label="' +
                getLocalize().flyout_onlineinfo +
                '"/> \n <MenuItem Label="---" /> <MenuItem Id="toolMock" Label="' +
                getLocalize().flyout_toolguide + 
                '" Checkable="true" Checked="false"/> \n<MenuItem Id="videoMock" Label="' +
                getLocalize().flyout_videoguide +
                '" Checkable="true" Checked="false"/> \n<MenuItem Label="---" /> \n<MenuItem Id="closePanel" Label="' +
                getLocalize().flyout_closepanel + 
                '" Checkable="true" Checked="false"/>\n</Menu>'
            )
        }

        // Uses the XML string to build the menu
        csInterface.setPanelFlyoutMenu(flyoutXML());

        // Flyout Menu Click Callback
        function toolGuideMenuClickedHandler(e) {

            // the event's "data" attribute is an object, which contains "menuId" and "menuName"
            console.dir(e);
            switch (e.data.menuId) {
                case "AboutItemAnimD2":
                    csInterface.evalScript("alert('AnimDessin2 \\nVersion 2.2.1\\nPhotoshop CC 2015 to 2018+\\n©2018 Stephane Baril\\nhttp://www.sbaril.me');");
                    break;
                case "OpenWebsiteAnimD2":
                    csInterface.openURLInDefaultBrowser("https://www.youtube.com/playlist?list=PLnNlOgl2T6GAsPqZDaf57TAqBa6ltL_gQ");
                    LoseFocus();
                    // csInterface.openURLInDefaultBrowser("http://www.sbaril.me/links"); LoseFocus();
                    break;
                case "videoMock":
                    toggleGuide("videoMock");
                    break;
                case "toolMock":
                    toggleGuide("toolMock");
                    break;
                case "closePanel":
                    toggleGuide("close");
                    break;
                default:
                    console.log(e.data.menuName + " clicked!");
            }
        }


        // Listen for the Flyout menu clicks
        csInterface.addEventListener("com.adobe.csxs.events.flyoutMenuClicked", toolGuideMenuClickedHandler);

        // Righclick Context Menu > uses exact same structure for now. easier acces
        function contextHandler(menuId) {
            switch (menuId) {
                case "AboutItemAnimD2":
                    csInterface.evalScript("alert('AnimDessin2 \\nVersion 2.2.1\\nPhotoshop CC 2015 to 2018+\\n©2018 Stephane Baril\\nhttp://www.sbaril.me');");
                    break;
                case "OpenWebsiteAnimD2":
                    csInterface.openURLInDefaultBrowser("https://www.youtube.com/playlist?list=PLnNlOgl2T6GAsPqZDaf57TAqBa6ltL_gQ");
                    LoseFocus();
                    // csInterface.openURLInDefaultBrowser("http://www.sbaril.me/links");
                    LoseFocus();
                    break;
                case "videoMock":
                    toggleGuide("videoMock");
                    break;
                case "toolMock":
                    toggleGuide("toolMock");
                    break;
                case "closePanel":
                    toggleGuide("close");
                    break;
            }
        }
        // Uses the XML string to build the menu
        // csInterface.setContextMenu(flyoutXML(), contextHandler);
        csInterface.setContextMenu(flyoutXML(), contextHandler);

    }
    init();

    //Dirty workarouned checking which page is loaded,
    // So context menu's can be checked
    if (document.title == getLocalize().PNL_Title_ToolGuide) {
        csInterface.updatePanelMenuItem(getLocalize().flyout_toolguide, true, true);
        csInterface.updatePanelMenuItem(getLocalize().flyout_videoguide, true, false);
        csInterface.updateContextMenuItem("toolMock", true, true);
        csInterface.updateContextMenuItem("videoMock", true, false);
    }
    if (document.title == getLocalize().PNL_Title_VideoGuide) {
        csInterface.updatePanelMenuItem(getLocalize().flyout_toolguide, true, false);
        csInterface.updatePanelMenuItem(getLocalize().flyout_videoguide, true, true);
        csInterface.updateContextMenuItem("toolMock", true, false);
        csInterface.updateContextMenuItem("videoMock", true, true);
    }
    console.log(document.title)
    // checkSettings();
    // getLocalize();
}());

com.sbaril.animdessin2/js/toolguides.js
/*jslint vars: true, plusplus: true, devel: true, nomen: true, regexp: true, indent: 4, maxerr: 50 */
/*global $, window, location, CSInterface, SystemPath, themeManager*/

(function () {
    'use strict';

    var csInterface = new CSInterface();
    var gExtensionId = "com.sbaril.animdessin2.Panel2";

    function Persistent(inOn) { // keep session stored

        if (inOn) {
            var event = new CSEvent("com.adobe.PhotoshopPersistent", "APPLICATION");
        } else {
            var event = new CSEvent("com.adobe.PhotoshopUnPersistent", "APPLICATION");
        }

        event.extensionId = gExtensionId;
        csInterface.dispatchEvent(event);
    }
    // Localize Languages
    function getLocalize() {
        var csInterface = new CSInterface();
        // csInterface.initResourceBundle();
        var renderBundle = csInterface.initResourceBundle();
        console.log(renderBundle)
        return renderBundle
    }

    // Opens the chrome developer tools in host app
    // Depricated https://community.adobe.com/t5/labs/showdevtools-method-added-by-default/td-p/6905305?page=1
    function showDevTools() {
        window.__adobe_cep__.showDevTools();
    }

    // Reloads extension panel
    function reloadPanel() {
        location.reload();
    }

    // Loads / executes a jsx file
    function loadJSXFile(pPath) {
        var scriptPath = csInterface.getSystemPath(SystemPath.EXTENSION) + pPath;
        csInterface.evalScript('evalFile("' + scriptPath + '")');
    }

    function checkSettings() {
        toggleGuide();
    }

    function callVideoMock() {
        window.location.href = "./videoguides.html";
    }
    function callToolMock() {
        window.location.href = "./toolguides.html";
    }
    // Show helpguide
    function toggleGuide(item) {
        if (item == "guide") {
            // else {
            localStorage.setItem('toggleGuide', "guideInActive");
            localStorage.setItem('toggleVideo', "videoInActive");
            $("#guide").addClass("InActive").removeClass("Active");
            $("#video").addClass("InActive").removeClass("Active");
            // document.getElementById("toggleVideo").Checked;
            // }
            if (localStorage.getItem('toggleGuide') == "guideInActive") {
                localStorage.setItem('toggleGuide', "guideActive");
                localStorage.setItem('toggleVideo', "videoInActive");
                $("#guide").addClass("Active").removeClass("InActive");
                $("#video").addClass("InActive").removeClass("Active");
                csInterface.resizeContent(400, 500);
                csInterface.updateContextMenuItem("toggleGuide", true, true);
                csInterface.updateContextMenuItem("toggleVideo", true, false);

            }
        }
        if (item == "close") {
            csInterface.closeExtension();
        }
        if (item == "videoMock") {
            csInterface.resizeContent(720, 480);
            callVideoMock();
        }
        if (item == "toolMock") {
            csInterface.resizeContent(400, 500);
            callToolMock();
        }
    }

    function sendWarning() {
        loadJSXFile("/jsx/AnimD2_warning.jsx");
    }

    function init() {
        themeManager.init();

        //Persistent(true); //persistent to prevent extension from unloading


        ////////////////////////////////////////////////////////////////////////////////
        // ANIMDESSIN2 TOOLGUIDE BUTTONS FUNCTIONS
        ////////////////////////////////////////////////////////////////////////////////   

        $("#btn_debug").click(showDevTools);
        $("#btn_reload").click(reloadPanel);

        $("#btn_closeGuide").click(toggleGuide);
        $("#btn_closeVideo").click(toggleGuide);



        ////////////////////////////////////////////////////////////////////////////////
        // ANIMDESSIN TOOLGUIDE FLYOUT MENU
        ////////////////////////////////////////////////////////////////////////////////

        // Ugly workaround to keep track of "checked" and "enabled" statuses
        var checkableMenuItem_isChecked = true;
        var targetMenuItem_isEnabled = true;


        function flyoutXML() {
            return "<Menu> \n<MenuItem Id=\"toolMock\" Label=\"" + (getLocalize().flyout_toolguides) + "\" Checkable=\"true\" Checked=\"false\"/> \n<MenuItem Id=\"videoMock\" Label=\"" + (getLocalize().flyout_videoguides) + "\" Checkable=\"true\" Checked=\"false\"/> \n <MenuItem Label=\"---\" /> \n<MenuItem Id=\"closePanel\" Label=\"" + (getLocalize().flyout_closepanel) + "\" Checkable=\"true\" Checked=\"false\"/> \n</Menu>";
        };

        // Uses the XML string to build the menu
        csInterface.setPanelFlyoutMenu(flyoutXML());

        // Flyout Menu Click Callback
        function toolGuideMenuClickedHandler(e) {

            // the event's "data" attribute is an object, which contains "menuId" and "menuName"
            console.dir(e);
            switch (e.data.menuId) {
                case "videoMock":
                    toggleGuide("videoMock");
                    break;
                case "toolMock":
                    toggleGuide("toolMock");
                    break;
                case "closePanel":
                    toggleGuide("close");
                    break;
                default:
                    console.log(e.data.menuName + " clicked!");
            }
        }


        // Listen for the Flyout menu clicks
        csInterface.addEventListener("com.adobe.csxs.events.flyoutMenuClicked", toolGuideMenuClickedHandler);

        // Righclick Context Menu > uses exact same structure for now. easier acces
        function contextHandler(menuId) {
            if (menuId == "toggleGuide") {
                toggleGuide("guide");
            }
            if (menuId == "videoMock") {
                toggleGuide("videoMock");
            }
            if (menuId == "toolMock") {
                toggleGuide("toolMock");
            }
            if (menuId == "closePanel") {
                toggleGuide("close");
            }
        }
        // Uses the XML string to build the menu
        csInterface.setContextMenu(flyoutXML(), contextHandler);

    }
    init();

    //Dirty workarouned checking which page is loaded,
    // So context menu's can be checked
    if (document.title == getLocalize().PNL_Title_ToolGuides) {
        csInterface.updatePanelMenuItem(getLocalize().flyout_toolguides, true, true);
        csInterface.updatePanelMenuItem(getLocalize().flyout_videoguides, true, false);
        csInterface.updateContextMenuItem("toolMock", true, true);
        csInterface.updateContextMenuItem("videoMock", true, false);
        console.log("banananaan")
    }
    if (document.title == getLocalize().PNL_Title_VideoGuides) {
        csInterface.updatePanelMenuItem(getLocalize().flyout_toolguides, true, false);
        csInterface.updatePanelMenuItem(getLocalize().flyout_videoguides, true, true);
        csInterface.updateContextMenuItem("toolMock", true, false);
        csInterface.updateContextMenuItem("videoMock", true, true);
        console.log("appels")
    }
    console.log(document.title)
    // console.log($('img').width());
    // checkSettings();
    // getLocalize();
}());

com.sbaril.animdessin2/js/globaltimeline.js
/*jslint vars: true, plusplus: true, devel: true, nomen: true, regexp: true, indent: 4, maxerr: 50 */
/*global $, window, location, CSInterface, SystemPath, themeManager*/

(function () {
    "use strict";

    var csInterface = new CSInterface();
    var gExtensionId = "com.sbaril.animdessin2.Timeline";

    function Persistent(inOn) {
        // keep session stored

        if (inOn) {
            var event = new CSEvent("com.adobe.PhotoshopPersistent", "APPLICATION");
        } else {
            var event = new CSEvent("com.adobe.PhotoshopUnPersistent", "APPLICATION");
        }

        event.extensionId = gExtensionId;
        csInterface.dispatchEvent(event);
    }

    // Opens the chrome developer tools in host app
    // Depricated https://community.adobe.com/t5/labs/showdevtools-method-added-by-default/td-p/6905305?page=1
    function showDevTools() {
        window.__adobe_cep__.showDevTools();
    }

    // Reloads extension panel
    function reloadPanel() {
        location.reload();
    }

    // Localize Languages
    function getLocalize() {
        var csInterface = new CSInterface();
        var renderBundle = csInterface.initResourceBundle();
        // console.log(renderBundle)
        return renderBundle
    }
 
    // Iconsize > done with fontsize
    function iconSize(size) {
        $("body").removeClass("iconSml").removeClass("iconMed").removeClass("iconBig").addClass(size);
        storeSettings(size);
        // document.getElementsByTagName('BODY').className = document.getElementsByTagName('BODY').className.replace(/(?:^|\s)\"+size+\"(?!\S)/g ,"");
    }

    // Set timecode > time or frames
    function timeCode(timecode) {
        storeSettings(timecode);
    }

    function checkSettings() {
        if (localStorage.getItem("timeCode")) {
            var timeCode = localStorage.getItem("timeCode");
            timeCode = timeCode == "false" ? false : true;
            localStorage.setItem("timeCode", timeCode);
            csInterface.updatePanelMenuItem(getLocalize().flyout_timeCode, true, timeCode);
            csInterface.updateContextMenuItem("timeCode", true, timeCode);
        }
        var openTimeline = "globalTimeline";
        $("body").addClass(openTimeline);
        $("#track").addClass(openTimeline);
    }

    //Store settings
    function storeSettings(store) {
        // alert(store)
        if (store == "onion") {
            if (localStorage.getItem("onionSkin") == "onionDisabled") {
                localStorage.setItem("onionSkin", "onionActive");
            } else {
                localStorage.setItem("onionSkin", "onionDisabled");
            }
        }
        if (store == "keys") {
            if (localStorage.getItem("keySettings") == "keysDisabled") {
                localStorage.setItem("keySettings", "keysActive");
            } else {
                localStorage.setItem("keySettings", "keysDisabled");
            }
        }
        if (store == "iconSml") {
            localStorage.setItem("iconSize", "iconSml");
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconSml, true, true);
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconMed, true, false);
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconBig, true, false);
            csInterface.updateContextMenuItem("iconSml", true, true);
            csInterface.updateContextMenuItem("iconMed", true, false);
            csInterface.updateContextMenuItem("iconBig", true, false);
        }
        if (store == "iconMed") {
            localStorage.setItem("iconSize", "iconMed");
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconSml, true, false);
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconMed, true, true);
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconBig, true, false);
            csInterface.updateContextMenuItem("iconSml", true, false);
            csInterface.updateContextMenuItem("iconMed", true, true);
            csInterface.updateContextMenuItem("iconBig", true, false);
        }
        if (store == "iconBig") {
            localStorage.setItem("iconSize", "iconBig");
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconSml, true, false);
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconMed, true, false);
            csInterface.updatePanelMenuItem(getLocalize().flyout_iconBig, true, true);
            csInterface.updateContextMenuItem("iconSml", true, false);
            csInterface.updateContextMenuItem("iconMed", true, false);
            csInterface.updateContextMenuItem("iconBig", true, true);
        }
        if (store == "timeCode") {
            var timeCode = localStorage.getItem("timeCode");
            timeCode = timeCode == "false" ? true : false;
            localStorage.setItem("timeCode", timeCode);
            csInterface.updatePanelMenuItem(getLocalize().flyout_timeCode, true, timeCode);
            csInterface.updateContextMenuItem("timeCode", true, timeCode);
        }
        console.log(store);
    }

    // Localize Languages
    function getLocalize() {
        var csInterface = new CSInterface();
        // csInterface.initResourceBundle();
        var renderBundle = csInterface.initResourceBundle();
        // console.log(renderBundle)
        return renderBundle;
    }

    function init() {
        themeManager.init();
        getLocalize();
        // Persistent(true); //persistent to prevent extension from unloading

        ////////////////////////////////////////////////////////////////////////////////
        // ANIMDESSIN2 CHECK FOR DOC
        ////////////////////////////////////////////////////////////////////////////////
        // checkOpenDoc();
        

        ////////////////////////////////////////////////////////////////////////////////
        // ANIMDESSIN FLYOUT MENU
        ////////////////////////////////////////////////////////////////////////////////

        // // Ugly workaround to keep track of "checked" and "enabled" statuses
        var checkableMenuItem_isChecked = true;
        var targetMenuItem_isEnabled = true;

        function flyoutXML() {
            return (
                '<Menu> \n<MenuItem Id="AboutItemAnimD2" Label="' +
                getLocalize().flyout_about +
                '"/> \n<MenuItem Id="OpenWebsiteAnimD2" Label="' +
                getLocalize().flyout_onlineinfo +
                '"/> \n <MenuItem Label="---" /> \n<MenuItem Id="iconSml" Label="' +
                getLocalize().flyout_iconSml +
                '" Checkable="true" Checked="false"/> \n\n<MenuItem Id="iconMed" Label="' +
                getLocalize().flyout_iconMed +
                '" Checkable="true" Checked="false"/> \n\n<MenuItem Id="iconBig" Label="' +
                getLocalize().flyout_iconBig +
                '" Checkable="true" Checked="false"/> \n\n<MenuItem Id="timeCode" Label="' +
                getLocalize().flyout_timeCode +
                '" Checkable="true" Checked="false"/> \n\n<MenuItem Label="---" /> \n <MenuItem Id="help" Label="' +
                getLocalize().flyout_helpguides +
                '" Checkable="true" Checked="false"/>\n<MenuItem Label="---" /> \n<MenuItem Id="closePanel" Label="' +
                getLocalize().flyout_closepanel + 
                '" Checkable="true" Checked="false"/>\n</Menu>'
            )
        }

        // Uses the XML string to build the menu
        csInterface.setPanelFlyoutMenu(flyoutXML());

        // Flyout Menu Click Callback
        function flyoutMenuClickedHandler(e) {
            // the event's "data" attribute is an object, which contains "menuId" and "menuName"
            console.dir(e);
            switch (e.data.menuId) {
                case "AboutItemAnimD2":
                    csInterface.evalScript("alert('AnimDessin2 \\nVersion 2.2.1\\nPhotoshop CC 2015 to 2018+\\n©2018 Stephane Baril\\nhttp://www.sbaril.me');");
                    break;
                case "OpenWebsiteAnimD2":
                    csInterface.openURLInDefaultBrowser("https://www.youtube.com/playlist?list=PLnNlOgl2T6GAsPqZDaf57TAqBa6ltL_gQ");
                    LoseFocus();
                    // csInterface.openURLInDefaultBrowser("http://www.sbaril.me/links"); LoseFocus();
                    break;
                case "iconSml":
                    iconSize("iconSml");
                    break;
                case "iconMed":
                    iconSize("iconMed");
                    break;
                case "iconBig":
                    iconSize("iconBig");
                    break;
                case "timeCode":
                    timeCode("timeCode");
                    break;
                case "help":
                    openGuides();
                    break;
                case "localize":
                    getLocalize();
                    break;
                case "closePanel":
                    toggleGuide("close");
                    break;
                default:
                    console.log(e.data.menuName + " panel clicked!");
            }
        }

        // Listen for the Flyout menu clicks
        csInterface.addEventListener("com.adobe.csxs.events.flyoutMenuClicked", flyoutMenuClickedHandler);

        // Righclick Context Menu > uses exact same structure for now. easier acces
        function contextHandler(menuId) {
            switch (menuId) {
                case "AboutItemAnimD2":
                    csInterface.evalScript("alert('AnimDessin2 \\nVersion 2.2.1\\nPhotoshop CC 2015 to 2018+\\n©2018 Stephane Baril\\nhttp://www.sbaril.me');");
                    break;
                case "OpenWebsiteAnimD2":
                    csInterface.openURLInDefaultBrowser("https://www.youtube.com/playlist?list=PLnNlOgl2T6GAsPqZDaf57TAqBa6ltL_gQ");
                    // csInterface.openURLInDefaultBrowser("http://www.sbaril.me/links");
                    LoseFocus();
                    break;
                case "iconSml":
                    iconSize("iconSml");
                    break;
                case "iconMed":
                    iconSize("iconMed");
                    break;
                case "iconBig":
                    iconSize("iconBig");
                    break;
                case "timeCode":
                    timeCode("timeCode");
                    break;
                case "help":
                    openGuides();
                    break;
                case "localize":
                    getLocalize();
                    break;
                case "closePanel":
                    toggleGuide("close");
                    break;
                default:
                    console.log(menuId + " panel clicked!");
            }
        }
        // Uses the XML string to build the menu
        csInterface.setContextMenu(flyoutXML(), contextHandler);
    }
    // $("#loaderBlock").hide();
    init();
    checkSettings();
})();
